### ISSUE TITLE
Pluggable architecture for impress.js

### ISSUE THREAD

**Issue by henrikingo** [impress/impress.js#494]
In some other issues, @bartaz has explained that he would like impress.js grow in functionality by the current core remaining much the same, but allowing for plugins that provide extended functionality. Lots of people have lots of ideas how they'd like to add to impress.js. In this writeup I've tried to collect all ideas (aka requirements) I've heard of, and then propose a high level architecture to meet those. In the end I've listed things that need to be discussed, aka I need help with those.

User stories
------------

I've heard people wanting to do following things with impress.js

 * I want to do presentations that are nicer than PowerPoint slides
   * I like writing presentations in HTML
   * I want to create a GUI tool for creating impress.js presentations
   * I need to be able to use this PowerPoint replacement without internet 
     connectivity
   * I want it to be something I can install as an application on my laptop
 * I want to write my presentations in some ohter markup language (MarkDown, 
   RST) and compile that into impress.js
 * I want to embed 3d animations in my web page or app
 * I want to embed impress.js functionality into my product (e.g. as a Drupal 
   module)
 * I want to extend the HTML notation used to create presentations (e.g. 
 relative positioning)
 * I want to add some standard visual elements to my presentations (e.g. 
 back & forward buttons)
 * I want impress.js to ship with a repository of nice templates I can base my 
 presentations on
 * I want to share my presentation to remote viewers over HTTP, while I control 
 which step they see
 
Note that implementations of most of the above already exists as separate 
projects or pull requests, which is how we know about them.
 
Architecture diagram
--------------------
 
We want to keep the core impress.js pretty much doing what it does today, but 
also create plugin APIs to allow for extensions that implement higher level
features without bloating the core. In fact, some of the above requirements
would be contradictory: a user wanting a big ass PowerPoint replacement wants
completely different functionality from someone wanting to embed a 3D animation
in his CMS software.

Taking the requirements above, we get an architecture like in the attached 
SVG diagram. Ahem, so github doesn't support SVG, here's a PNG, I'll try to share the original somewhere later.

![impress-architecture-diagram](https://cloud.githubusercontent.com/assets/2482018/9831089/bf3a713a-5951-11e5-9fa5-b13f995885f0.png)

For discussion
--------------

Following decisions need to be made by a person more familiar with current
state of art in JavaScript than myself:

 * How are plugins and APIs created (coding style, how are plugins loaded, 
   dependency injection...)
 * Repositories: Is every plugin a separate github repository? If yes, how 
   are they brought together and version dependencies resolved?
 * Packaging: bower, npm, rpm? Note that different user stories may lead to 
   different package repositories (e.g. bower for a HTML5 app, RPM for a 
   PowerPoint replacement)
 * Lower level details like namespace conventions (how to avoid 2 plugins using
   same names in JS, CSS, HTML attributes)

**Comment by FagnerMartinsBrack** [comment]
I am not the best authority to provide the answers, but I am familiar enough with JS ecosystem to answer these questions.

> How are plugins and APIs created (coding style, how are plugins loaded, dependency injection...)

Code style is subjetive, use a preset from [jscs](https://github.com/jscs-dev/node-jscs/tree/master/presets) to prevent [bikeshedding](https://en.wikipedia.org/wiki/Parkinson%27s_law_of_triviality).

Dependency injection I only saw in Angular, JS have no native dependency injection mechanism

Plugins can be loaded in several ways, usually people trigger events with the pub/sub pattern. Jmpress.js [has a very good subscribe mechanism](https://github.com/jmpressjs/jmpress.js/blob/master/src/plugins/duration.js), but I am not sure how to adapt to impress.js from the top of my head.

> Repositories: Is every plugin a separate github repository?

Not if it is essential. Take jquery for example, they changed sizzle (the selectors library) to another repository but initally it was built in. In impress.js case I would recommend creating a robust plugin mechanism and create plugins, like keyboard navigation, in a separate repository. There is also the possibility to create repositories for specific use-cases like "impress.js as a common presentation", "impress.js as a site navigator", although that doesn't seems appealing.

> If yes, how are they brought together and version dependencies resolved?

Usually we use npm or bower and integrate in a build mechanism using gulp or grunt. If we want a new core version it's just a matter of updating the dependency version in the `package.json` file.

> Packaging: bower, npm, rpm?

The 2 most used are bower and npm. bower is fading away slowly, but npm is here to stay.

> Note that different user stories may lead to different package repositories (e.g. bower for a HTML5 app, RPM for a PowerPoint replacement)

I don't see how that should be responsibility of the org, other than provide the code. One package manager should be enough, and then others could be added if there's A LOT of requests and people available.

> Lower level details like namespace conventions (how to avoid 2 plugins using same names in JS, CSS, HTML attributes)

There is no way to prevent 2 libs using the same namespace, what we do when we know that there is a conflicting namespace is to use a [noConflict pattern](https://github.com/js-cookie/js-cookie/pull/24), jQuery [does the same](https://github.com/jquery/jquery/blob/c9cf250daafe806818da1dd207a88a8e94a4ad16/src/exports/global.js#L9) (in their way).

For plugins it's just assuming that people will "own" their namespace inside impress.js at the moment they create one, although the namespace priority will be from impress.js dev itself. Take for example Grunt, they call "grint-contrib-*" all plugins that are maintained by the Grunt team, something similar could be done with impress.js.

I hope it helps.

**Comment by henrikingo** [comment]
Cool. Progress within a single day!

I suppose I should rephrase the point about dependencies then: maybe dependency injection is not needed, but certainly there must be a way to specify that "this module won't work unless module B is loaded first". Of course, if npm is used, it takes care of a large part of that problem.

I'll have to read up on your links, thanks a lot!

**Comment by FagnerMartinsBrack** [comment]
> ... there must be a way to specify that "this module won't work unless module B is loaded first"

Just docs, JS is untyped, In jQuery case they just document that "this is a jQuery plugin" and the dev is responsible to load jQuery first.

**Comment by henrikingo** [comment]
Hi @FagnerMartinsBrack 

It seems I forgot to ask for input on test frameworks. Is Mocha the way to go also for client side JS? Do you have suggestions to look at anything else?

**Comment by FagnerMartinsBrack** [comment]
Mocha, QUnit. There is a lot of possibilities. I personally use QUnit integrated with Travis CI Open-Source plan (Continuous Integration) integrated with Sauce Labs Open-Source plan (Cross-browser testing).

I have worked (although didn't set up from scratch) with a project using Karma (it starts a local browser to run tests and allow you to debug it). I would say Karma is the best for crossbrowser testing and debugging, but didn't have any chances to work fully on it (yet). You can use QUnit, Mocha or Kasmine to run tests on top of it (Karma is not a test framework, it just start the browsers).

**Comment by FagnerMartinsBrack** [comment]
For assertions I am most familiar with QUnit, that was created and is in use by jQuery Team. Karma is used and was built by the AngularJS Team.

**Comment by LongLiveCHIEF** [comment]
I was thinking of doing this:

- creating an impress.js IDE/GUI using [electron](atom.electron.io)
- using raphael.js for animations
- using fabric.js for SVG generation/import, freehand drawing, etc...
- using jsgit internally for revision history (undo/redo)

Overall, the experience within the IDE should be familiar to users of PowerPoint/Keynote/etc..., but produce amazing impress.js presentations.

**Comment by henrikingo** [comment]
I thought I should have a lot of free time during the Autumn, turns out I was wrong. But during Winter I've been able to take a stab at adding a test framework (chose qunit) and wrote some basic tests to get started. As you can see above, there's now a pull request.

If we can get this merged, I invite others to write more tests, in particular:
 - the keypress api at the end of impress.js
 - testing that data-* attributes convert correctly to CSS, zooming, positioning, animations...

PS. @LongLiveCHIEF that sounds awesome, and also like a lot of work!

**Comment by henrikingo** [comment]
I wanted to note that I've now made progress with a proposal for a plugin api. Please see #541 and comment there if you want to discuss.

**Comment by henrikingo** [comment]
Looks like we've made some good progress now on laying a foundation for a more feature rich impress.js.
* qunit test framework is now merged, although test coverage in master is still shallow.
* I have a repo with a plugin api, where I've already implemented several new plugins.
* There are bower and npm json files in the repo as well, but I'm unsure if they're correct/current. Anyway, I consider packaging to be a somewhat organic function: if someone needs some packaging format, it tends to emerge.

As a conclusion, it is time to close this ticket. Thanks for helping out.
